internal/compiler/passes/resolving/remove_noop.rs

```diff
@@ -0,0 +1,110 @@
+        // effects.
+        //
+        // Remove all expressions without side effects (except for the last one) and emit a
+        // warning.
+        //
+        // Note: Iterate over the indices in reverse, so that all to-be-iterated indices remain
+        // valid when removing items from the vector.
+        for index in (0..(code_block.len() - 1)).rev() {
+            let (node, expression) = &code_block[index];
+            if without_side_effects(expression) {
+                diagnostics.push_warning("Expresssion has no effect!".to_owned(), node);
+                code_block.remove(index);
+            }
+        }
+    }
+}
+
+/// Returns whether the expression is certain to be without side effects.
+/// This function is conservative and may still return `false`, even if a given expression
+/// is without side effects.
+/// It is only guaranteed that if this function returns `true`, the expression definitely does not
+/// contain side effects.
+fn without_side_effects(expression: &Expression) -> bool {
+    match expression {
+        Expression::Condition { condition, true_expr, false_expr } => {
+            without_side_effects(condition)
+                && without_side_effects(true_expr)
+                && without_side_effects(false_expr)
+        }
+        Expression::NumberLiteral(_, _) => true,
+        Expression::StringLiteral(_) => true,
+        Expression::BoolLiteral(_) => true,
+        Expression::CodeBlock(expressions) => expressions.iter().all(without_side_effects),
+        Expression::FunctionParameterReference { .. } => true,
+        // Invalid and uncompiled expressions are unknown at this point, so default to
+        // `false`, because they may have side-efffects.
+        Expression::Invalid => false,
+        Expression::Uncompiled(_) => false,
+        // A property reference may cause re-evaluation of a property, which may result in
+        // side effects
+        Expression::PropertyReference(_) => false,
+        Expression::ElementReference(_) => false,
+        Expression::RepeaterIndexReference { .. } => true,
+        Expression::RepeaterModelReference { .. } => true,
+        Expression::StoreLocalVariable { .. } => false,
+        Expression::ReadLocalVariable { .. } => true,
+        Expression::StructFieldAccess { base, name: _ } => without_side_effects(&*base),
+        Expression::ArrayIndex { array, index } => {
+            without_side_effects(&*array) && without_side_effects(&*index)
+        }
+        // Note: This assumes that the cast itself does not have any side effects, which may not be
+        // the case if custom casting rules are implemented.
+        Expression::Cast { from, to: _ } => without_side_effects(from),
+        // Note: Calling a *pure* function is without side effects, however
+        // just from the expression, the purity of the function is not known.
+        // We would need to resolve the function to determine its purity.
```

Comment:
one could run this pass later. But I understand this is done because otherwise we no longer have the location. But perhaps we could add it just to CodeBlock.
But this is fine for now.